### DISCUSSION TITLE
validation of input values and error handling at useAsyncData()

### DISCUSSION THREAD

**Discussion by naoyoshi** [nuxt/nuxt#18067] [Questions] [answered]
According to https://v3.nuxtjs.org/docs/usage/data-fetching, useAsyncData() method returns refresh and error.
I have 2 questions about validation of input value and error handling.

1)validation of input value
I examined to work function of refresh, then it seems that it can't validate input values inside function of refresh.
Could anyone validate values?
If anyone could, please tell me the sample code?

2)error handling
Now I can't validate input values before request to a server side, I thought that validation of values is in server side.
Then if the value has error, I code to return as 422.
But I don't know how to handle this error object at useAsyncData().
Could anyone handle?
If anyone could, please tell me the sample code?

**Comment by viii47** [comment, marked as answer]
This is relatively straightforward in the setup lifecycle. You need to do *something* with the refresh function -> either you call it in the template directly, you expose it to the `method()` function with `this.refresh()` rather or you simply watch your desired values, functions, inputs, etc.

Notes to this:
- `metawatchInputFeedback` is an exposed watch function, purposely so we can kill it when the component/page/etc. is unmounted (important lifecycle -> otherwise it might still be called). Calling the function by itself will suffice to stop it (see `beforeUnmount() -> this.metawatchInputFeedback()`)
- useAsyncData() needs your input values inside the `headers: { ... }` - How you define and update them is up to you. I would simply have an exposed collection of an Object that I'd update via a method handler
-  the watch function `const metawatchInputFeedback = ...` first property is what you would like to watch. Anything works here - route, router, collection of values, a simple boolean variable and more


```js
export default defineComponent({
    async setup() {
        const route = useRoute();
        const { data: inputfeedback, refresh } = await useAsyncData('YOURUNIQUEKEY', () =>
            $fetch(`${API_ENDPOINT}`, {
                method: 'POST',
                headers: {
                    authorization: 'Bearer',
                    input_values: {...},
                }
            })
        );

        const metawatchInputFeedback = watch(
            () => route.query.q, () => {
                refresh().finally(() => {
                     // Here you can handle whatever happens after the refresh is done
                    console.log('yay')
                });
            }
        )

        return {
            inputfeedback,
            refresh,
            metawatchInputFeedback
        }
    },
    beforeUnmount() {
        this.metawatchInputFeedback()
    }
})
```

**Reply by naoyoshi** [reply]
Thank you for your code.

<script setup> is convenient, but I need to pay attention when we should use <script setup> properly.
